### PR TITLE
[DROOLS-7253] Review and improve drools-lsp parser Visitor design

### PR DESCRIPTION
**Pull request by tkobayas**
- Remove currentConstructStack. Utilize return object instead.

### CODE REVIEW

**Comment by tkobayas** [comment]
Removed `currentConstructStack`. Instead, utilize return objects of Visitor methods, so it's more natural to AST traversing.

File: drools-parser/src/main/java/org/drools/parser/DRLVisitorImpl.java [hunk 1]
```diff
@@ -125,80 +118,98 @@ public Object visitRuledef(DRLParser.RuledefContext ctx) {
...
-        currentPattern = new PatternDescr(ctx.objectType.getText());
+    public PatternDescr visitLhsPattern(DRLParser.LhsPatternContext ctx) {
+        PatternDescr patternDescr = new PatternDescr(ctx.objectType.getText());
+        currentPattern = patternDescr;
         if (ctx.patternSource() != null) {
             String expression = ctx.patternSource().getText();
             FromDescr from = new FromDescr();
             from.setDataSource(new MVELExprDescr(expression));
             from.setResource(currentPattern.getResource());
             currentPattern.setSource(from);
```
Comment (gitgabrio) [inline drools-parser/src/main/java/org/drools/parser/DRLVisitorImpl.java:187]: Hi @tkobayas !
I'm not sure to follow the logic of `currentPattern.setSource(from);` (@187) and `currentPattern = null;` (@190) and, in general, of this method: at line @101 `currentPattern` points to same object as `patternDescr`; in the `if` statement the status of that object is changed (through `currentPattern`), and then `currentPattern`reference is nullified: I have the impression that that `currentPattern` is not needed here and, in general, being it a "status" of the class, make the code slightly harder to understand/maintain then needed: wdyt ?

**Review by gitgabrio — CHANGES_REQUESTED** [review]
HI @tkobayas 
Thanks for the PR: IMO it is better, but still I would avoid those three instance variables `packageDescr,` `currentRule` and `currentPattern`: as far as I see they are strictly related to the status of different "contexts", so, if possible, I think it would be better to embed them in the appropriate context, and altogether avoid to rely upon such "spurious" variables: wdyt ?

**Comment by tkobayas** [comment]
Removed `packageDescr`, `currentRule` and `currentPattern`. Now Descrs are constructed by return objects along with Visitor AST travseral. Could you share any thoughts on this? @gitgabrio @mariofusco

**Review by gitgabrio — APPROVED** [review]
Hi @tkobayas 
Thanks, you really imrpoved that! You get rid of all those hidden side effects and also proveded a better typing of returned object: LGTM!

**Review by mariofusco — APPROVED** [review]

**Pull request merged**